pulp_container/app/migrations/0023_data_translate_perms_to_roles.py

```diff
@@ -0,0 +1,241 @@
+            groups_to_delete.clear()
+
+    # Persist leftovers
+    UserRole.objects.bulk_create(user_roles)
+    Group.objects.filter(pk__in=groups_to_delete).delete()
+
+
+def translate_permissions_to_roles(apps, schema_editor):
+    AccessPolicy = apps.get_model("core", "AccessPolicy")
+    if any(AccessPolicy.objects.filter(viewset_name__in=CONTAINER_VIEWSET_NAMES).values_list("customized", flat=True)):
+        print("Container access policies have been modified. Translating the assigned permissions to roles will not be attemted.")
+        return
+
+    ContentType = apps.get_model("contenttypes", "ContentType")
+    User = apps.get_model(settings.AUTH_USER_MODEL)
+
+    Group = apps.get_model("core", "Group")
+    Role = apps.get_model("core", "Role")
+    UserRole = apps.get_model("core", "UserRole")
+    GroupRole = apps.get_model("core", "GroupRole")
+    Permission = apps.get_model("auth", "Permission")
+    UserPermission = apps.get_model(User.user_permissions.through._meta.app_label, User.user_permissions.through._meta.model_name)
+    GroupPermission = apps.get_model("auth", "Group_permissions")
+    UserObjectPermission = apps.get_model("guardian", "UserObjectPermission")
+    GroupObjectPermission = apps.get_model("guardian", "GroupObjectPermission")
+    ContainerNamespace = apps.get_model("container", "ContainerNamespace")
+    ContainerDistribution = apps.get_model("container", "ContainerDistribution")
+
+    user_roles = []
+    group_roles = []
+
+    # Find global permissions and transform into user/group roles
+    for app_label, codename, role_name in GLOBAL_PERMISSION_ROLE_TRANSLATOR:
+        perm = Permission.objects.filter(content_type__app_label=app_label, codename=codename).first()
+        if perm:
+            role, _ = Role.objects.get_or_create(name=role_name, defaults={"locked": True})
+            user_permissions = UserPermission.objects.filter(permission=perm)
+            user_roles.extend(UserRole(user=user_permission.user, role=role) for user_permission in user_permissions)
+            user_permissions.delete()
+            group_permissions = GroupPermission.objects.filter(permission=perm)
+            group_roles.extend(GroupRole(group=group_permission.group, role=role) for group_permission in group_permissions)
+            group_permissions.delete()
+
+            # Handle batches
+            if len(user_roles) > 1000:
+                UserRole.objects.bulk_create(user_roles)
+                user_roles.clear()
+            if len(group_roles) > 1000:
+                GroupRole.objects.bulk_create(group_roles)
+                group_roles.clear()
+
+    # Find object permissions and transform into user/group roles
+    for permission_names, rolename in OBJECT_PERMISSION_ROLE_TRANSLATOR:
+        perms = [Permission.objects.filter(content_type__app_label=app_label, codename=codename).first() for app_label, codename in permission_names]
+        if all(perms):
```

Comment:
@mdellweg, is my understanding correct, that it's the best we can do and there can be individual object permissions left not translated to the roles? The case when `all` returns  `False`.